### PR TITLE
oop-object_interactions-hard-q4

### PR DESCRIPTION
**Pull request by SamamthaH**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/object_interactions/hard/q4/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+- Java question
+
+## Question Text
+
+Some data should be encrypted, however to do so you need a key to base the encryption off of. Design part of a `CryptBox`
+class using the provided class components, and create a `crossEncrypt` method that takes another `CryptBox` object and 
+uses that object's key to encrypt the other object's data.
+
+### CryptBox Class:
+
```
Comment (KaiKhademi) [inline src/oop/object_interactions/hard/q4/QuestionText.md:12]: This is a hard level assignment so I guess they can write the whole object. If you want them to write it partially their tests should be in setup(). But here its better to just ask them to write it all.

File: src/oop/object_interactions/hard/q4/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    - Both of these attributes should not be accessible from another class
+
+- Constructor:
+    - A constructor visible from another class that accepts a `String` and an `int` parameter, and initializes the 
+      **lockedData** attribute with the `String` and the **cryptKey** attribute with the `int`
+    - This constructor should be visible from another class
+
+- Methods
+    - A `getCryptKey` method which returns the **cryptKey** attribute of the `CryptBox`
+
```
Comment (KaiKhademi) [inline src/oop/object_interactions/hard/q4/QuestionText.md:27]: Maybe change the visibility modifier wording of the items that are already given to "is visible" or "is not accessible" rather the "should be visible" or "should be accessible"

File: src/oop/object_interactions/hard/q4/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void correctGetCryptKeyMethod(String data, int key) throws Throwable {
+        Object[][] constructorArguments = {
+                {data, String.class},
+                {key, int.class}
+        };
+        Object instance = classInstance.createInstance(constructorArguments);
+        String incorrectGetMethods = "Your %s method does not return the value of the %s attribute.";
+        assertTrue(classInstance.hasMethod(getIntAttribute1, null, int.class, new String[]{"public"}),
```
Comment (KaiKhademi) [inline src/oop/object_interactions/hard/q4/MainTest.java:86]: I believe this is redundant as the callMethod and assertion check all these

**Review by KaiKhademi — CHANGES_REQUESTED** [review]

File: src/oop/object_interactions/hard/q4/QuestionText.md (hunk 1 above)
Comment (SamamthaH) [reply src/oop/object_interactions/hard/q4/QuestionText.md:12]: due to meeting, switching this to parsons instead

File: src/oop/object_interactions/hard/q4/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void cryptBoxClassHasCorrectConstructor(String data, int key) throws Throwable {
+        Object[][] arguments = {
+                {data, String.class},
+                {key, int.class}
+        };
+        assertTrue(classInstance.hasConstructor(new Class[]{String.class, int.class}),
+                String.format("Your %s class is missing a required constructor.", objectClassName));
+        assertTrue(classInstance.hasConstructor(new Class[]{String.class, int.class}, new String[]{"public"}),
```
Comment (KaiKhademi) [inline src/oop/object_interactions/hard/q4/MainTest.java:74]: You don't need these two here you are already testing them

**Review by KaiKhademi — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**